git: include shallow references in object walker

erka opened the pull request:
When walking objects in a repository, the objectWalker should check shallow references in addition to regular refs. This ensures that shallow clones are properly handled during operations like pruning or repacking.

Similar to #1785

erka committed:
git: include shallow references in object walker

When walking objects in a repository, the objectWalker should check
shallow references in addition to regular refs. This ensures that shallow
clones are properly handled during operations like pruning or repacking.

Signed-off-by: Roman Dmytrenko <rdmytrenko@gmail.com>

pjbgf commented on object_walker.go:
```diff
@@ -72,6 +74,13 @@ func (p *objectWalker) walkObjectTree(hash plumbing.Hash) error {
 		for _, h := range obj.ParentHashes {
 			err = p.walkObjectTree(h)
 			if err != nil {
+				shallow, serr := p.Storer.Shallow()
```
This should be gated with `errors.Is` for `plumbing.ErrObjectNotFound`.

pjbgf commented on object_walker.go:
```suggestion
		return fmt.Errorf("getting object %s failed: %w", hash, err)
```

pjbgf commented on object_walker_test.go:
```diff
@@ -0,0 +1,82 @@
+	suite.Run(t, new(objectWalkerSuite))
+}
+
+func (s *objectWalkerSuite) TestNormalClonedRepo() {
```
Add another test to ensure that other errors are not silently ignored (e.g. I/O, disk permissions, etc).

pjbgf reviewed (changes requested):
@erka thanks for proposing this fix. I added some comments above, let me know if anything isn't clear.

erka committed:
Update object_walker.go

Co-authored-by: Paulo Gomes <paulo.gomes.uk@gmail.com>
Signed-off-by: Roman Dmytrenko <rdmytrenko@gmail.com>

erka committed:
address PR feedback

Signed-off-by: Roman Dmytrenko <rdmytrenko@gmail.com>

erka commented:
Thanks for your review @pjbgf. I've addressed your points.

pjbgf committed:
git: skip parents of shallow roots in object walker

Replace the error-path shallow detection with an upfront check
against a cached shallow set. Classifying shallow boundaries after
a parent walk failed swallowed unrelated errors: any error other
than ErrObjectNotFound surfacing from a parent subtree was silently
dropped for every repository, leaving reachable objects unmarked
and eligible for deletion by Prune.

Skipping the parent loop for commits listed in .git/shallow mirrors
git's graft semantics, avoids re-reading the shallow file on every
recursion frame, and restores unconditional error propagation in
the walker.

Signed-off-by: Paulo Gomes <paulo@entire.io>

pjbgf reviewed (dismissed):
@erka thanks for working on this. 🙇 

ℹ️ I added an additional commit to fix two things:
- The regression that replaced `err != nil` with the check for object not found. We needed both workflows, so that was resulting in some errors being ignored.
- Started front-loading the shallow list and verification. That is needed as currently there is no caching in `.Shallow()`.

pjbgf committed:
*: Fix linting issue

Signed-off-by: Paulo Gomes <paulo@entire.io>
